server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java

```diff
@@ -497,7 +500,8 @@ private GroupShardsIterator<SearchShardIterator> getIterator(
     ) {
         int cardinality = results.getNumPossibleMatches();
         FixedBitSet possibleMatches = results.getPossibleMatches();
-        if (cardinality == 0) {
+        // TODO: pick the local shard when possible
+        if (requireAtLeastOneMatch && cardinality == 0) {
```

Comment:
The new flag is only used here. I can make it in a separate PR, but the newly added tests will need to account this.

Reply:
thanks for raising this. What does the new flag help with?

Reply:
This flag has been introduced for the search_shards API. Currently, the CanMatchPreFilterSearchPhase ensures that at least one matching shard is returned, enabling us to generate the search response. I will investigate ways to remove this limitation. When the search_shards API marks a skipped shard as matched, that shard will never be skipped by the can_match phase.